crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1188,266 @@ struct Peer {
 ///
 /// This will keep track of unique transaction hashes that are currently being fetched and submits
 /// new requests on announced hashes.
-#[derive(Debug, Default)]
+#[derive(Debug)]
+#[pin_project]
 struct TransactionFetcher {
+    /// All peers to which a request for pooled transactions is currently active. Maps 1-1 to
+    /// `inflight_requests`.
+    active_peers: LruMap<PeerId, u8>,
     /// All currently active requests for pooled transactions.
+    #[pin]
     inflight_requests: FuturesUnordered<GetPooledTxRequestFut>,
-    /// Set that tracks all hashes that are currently being fetched.
-    inflight_hash_to_fallback_peers: HashMap<TxHash, Vec<PeerId>>,
+    /// Hashes that are awaiting fetch from an idle peer.
+    buffered_hashes: LruCache<TxHash>,
```

Comment:
we're likely opening us up to a DOS attack here, because this has a fixed capacity and is updated on pooled_tx broadcasts, peers could flood us with fake hashes that evict valid buffered ones?

Reply:
yeah that could happen. this is the flow outlined:
an announcement comes
announcement is filtered
any new hashes (new is defined as unknown and not in `unknown_hashes`, i.e. not buffered and not inflight) are packed in request for peer
if not enough space in request for all new hashes, hashes are buffered

...since many legit peers will announce the same tx hashes to us, it's quite unlikely valid tx hashes will be buffered. one way to get around the problem you describe is to increase the concurrency per peer, so that we can have 3 inflight requests per peer for 3 different hashes concurrently. that way no hashes will be buffered from reception of an announcement. then hashes will only buffered if the request failed.